### PR TITLE
Coding challenge

### PR DESCRIPTION
**Pull request by sharapat**
changed o'->ó, a'->á, u'->ú, g'->ǵ, n'->ń and ts->c, sh->ch where necessary

### CODE REVIEW

**Comment by IlnarSelimcan** [comment]
## Coverage on the [story](https://github.com/taruen/apertiumpp/blob/master/data4apertium/corpora/jam/kaa.txt) before @sharapat 's changes:

```
apertium-kaa$ git log -1
commit 54b29b3c1040fb45ead0155daa5ca3487ac934b5 (HEAD -> master, origin/master, origin/HEAD)
Author: Ilnar Salimzianov <xxxxxxxxxxxxxx>
Date:   Sun Apr 7 01:15:30 2019 +0300

    add .travis.yml and test.py

apertium-kaa$ aq-covtest ../../../data4apertium/corpora/jam/kaa.txt kaa.automorf.bin 
Number of tokenised words in the corpus: 489
Coverage: 69.53%
Top unknown words in the corpus:
17	 Ayg
13	 l
13	 k
7	 j
5	 h
4	 redi
4	 ı
4	 shigi
3	 lken
3	 K
3	 ldi
3	 pıshı
2	 andı
2	 zir
2	 shik
2	 Yaq
2	 yde
2	 n
2	 m
2	 qayerge
Translation time: 0.016431331634521484 seconds
apertium-kaa$ 
```

## Coverage on the story after sharapat's changes:

```
apertium-kaa$ git log -1
commit e7429c8f63c9cdb443ee0ffc0f8d897a64f47209 (HEAD -> coding-challenge, origin/coding-challenge)
Author: sharapat <xxxxxxxxxxxxxxxxxxxxx>
Date:   Wed Apr 24 00:26:03 2019 +0500

    changed sh -> ch where neccessary

apertium-kaa$ aq-covtest ~/1Working/1Taruen/apertiumpp/data4apertium/corpora/jam/kaa.txt kaa.automorf.bin 
Number of tokenised words in the corpus: 413
Coverage: 87.89%
Top unknown words in the corpus:
4	 kóredi
3	 júda
3	 Aygúldiń
3	 pıshıǵı
2	 oynaǵandı
2	 Yaq
2	 úyde
2	 qayerge
1	 barqulla
1	 úydiń
1	 aldındaǵı
1	 oynaǵanına
1	 baratır
1	 Aygúlden
1	 kózlerin
1	 kórip
1	 turǵanı
1	 bolǵannan
1	 atirapına
1	 kórdińizbe
Translation time: 0.019109010696411133 seconds
```
Unsurprisingly, the coverage has increased, which is a good thing.

@sharapat , this is how you can install aq-covtest:  https://gist.github.com/IlnarSelimcan/87efbbe8a569833e68e89b126836e2bc . It is a tool for measuring the coverage of a transducer, i.e. it shows percentage of words in a corpus the transducer recognizes.

As the next step, please check whether the *stems* of unknown words are in the `apertium-kaa.kaa.lexc`. In other words, make sure that `kór:kór V-TV ;` is there, and `Aygúl:Aygúl NP-ANT-F ;` is, etc.

If all stems from the story are already in the lexicon, it'll be easier for @jonorthwash to work on `apertium-kaa.kaa.twol` rules, since he'll know that unknown words `aq-covtest` returns from the story are unknown because of issues with twol, and *not* because their stems are missing from `apertium-kaa.kaa.lexc`.

@sharapat ,   also, we'll need some more texts written in (or converted to) the new orthography (for testing purposes).

These can be Karakalpak texts which you'd like to see translated to Uzbek (personal interest in translations is a good motivator). Any text works, but ideally they should be under a free license (public domain, creative commons-by, creative-commons-by-sa).

File: apertium-kaa.kaa.lexc [hunk 1]
```diff
@@ -7452,22 +7452,22 @@ mu%'mkin:mu%'mkin INTERJ ; ! "probably"
...
 km²:km2 ABBR ; ! "km²" Dir/LR
 km³:km³ ABBR ; ! "km³"
 km³:km3 ABBR ; ! "km³" Dir/LR
-ma%'ml:ma%'ml ABBR ; ! "ma'mleket"
+máml:máml ABBR ; ! "ma'mleket"
 t%.b%.:t%.b%. ABBR ; ! "etc."
 j%.:j%. ABBR ; ! "year"
 jj%.:jj%. ABBR ; ! "year"
-JIO%'.:JIO%' ABBR ; ! "GDP"
+JIó.:JIó ABBR ; ! "GDP"
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: `ó` should be capital letter.

File: apertium-kaa.kaa.lexc [hunk 2]
```diff
@@ -4449,29 +4448,29 @@ Xikmet:Xikmet NP-ANT-M ; ! "Xiykmet" (Arabic)
...
-Ma%'jitov:Ma%'jitov NP-COG-OV ; ! ""
-Da%'wqaraev:Da%'wqaraev NP-COG-OV ; ! ""
-Da%'ribaev:Da%'ribaev NP-COG-OV ; ! ""
-A%'wezov:A%'wezov NP-COG-OV ; ! ""
-Ma%'tmuratov:Ma%'tmuratov NP-COG-OV ; ! ""
-Ta%'jimuratov:Ta%'jimuratov NP-COG-OV ; ! ""
+Májitov:Májitov NP-COG-OV ; ! ""
+Dáwqaraev:Dáwqaraev NP-COG-OV ; ! ""
+Dáribaev:Dáribaev NP-COG-OV ; ! ""
+áwezov:áwezov NP-COG-OV ; ! ""
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: `á` should be capital.

File: apertium-kaa.kaa.lexc [hunk 3]
```diff
@@ -4449,29 +4448,29 @@ Xikmet:Xikmet NP-ANT-M ; ! "Xiykmet" (Arabic)
...
+Shınargúl:Shınargúl NP-ANT-F ; ! "Shınargül" (Arabic)
 Shınar:Shınar NP-ANT-F ; ! "Shınar" (Kazakh)
-Shın%'g%'ısxan:Shın%'g%'ısxan NP-ANT-M ; ! "" Mosty it%'s written separately, but occures as one word also
+Shıńǵısxan:Shıńǵısxan NP-ANT-M ; ! "" Mosty it%'s written separately, but occures as one word also
 Ela:Ela NP-ANT-F ; ! "Ela" (Latin)
 Era:Era NP-ANT-F ; ! "Era" (New word)
 İysa:İysa NP-ANT-M ; ! "Jesus"
 
-O%'tepov:O%'tepov NP-COG-OV ; ! ""
+ótepov:ótepov NP-COG-OV ; ! ""
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: capital

File: apertium-kaa.kaa.lexc [hunk 4]
```diff
@@ -4337,36 +4336,36 @@ Arna:Arna NP-ANT-F ; ! "Arna" (Kazakh)
...
-A%'bdiraxman:A%'bdiraxman NP-ANT-M ; ! "Äbdiraxman" (Arabic)
-A%'bil:A%'bil NP-ANT-M ; ! "Äbil" (Arabic)
-A%'bilqasım:A%'bilqasım NP-ANT-M ; ! "Äbilqasım" (Arabic)
-A%'bu:A%'bu NP-ANT-M ; ! "Äbiw" (Arabic)
-A%'di:A%'di NP-ANT-M ; ! "Ädiy" (Arabic)
-A%'dil:A%'dil NP-ANT-M ; ! "Ädil" (Arabic)
-A%'kim:A%'kim NP-ANT-M ; ! "Äkim" (Arabic)
-A%'lim:A%'lim NP-ANT-M ; ! "Älim" (Arabic)
-A%'mir:A%'mir NP-ANT-M ; ! "Ämir" (Kazakh)
+ábdiraxman:ábdiraxman NP-ANT-M ; ! "Äbdiraxman" (Arabic)
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: All names below should start with capital letter.

File: apertium-kaa.kaa.lexc [hunk 5]
```diff
@@ -4156,20 +4155,20 @@ Amerika:Amerika NP-TOP ; ! ""
...
 Zimbabve:Zimbabve NP-TOP ; ! ""
-Ta%'jikistan:Ta%'jikistan NP-TOP ; ! ""
+Tájikistan:Tájikistan NP-TOP ; ! ""
 Tuva:Tuva NP-TOP ; ! ""
 Pakistan:Pakistan NP-TOP ; ! ""
 Indiya:Indiya NP-TOP ; ! ""
-Tu%'rkmenistan:Tu%'rkmenistan NP-TOP ; ! ""
-A%'zerbayjan:A%'zerbayjan NP-TOP ; ! ""
+Túrkmenistan:Túrkmenistan NP-TOP ; ! ""
+ázerbayjan:ázerbayjan NP-TOP ; ! ""
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: capital

File: apertium-kaa.kaa.lexc [hunk 6]
```diff
@@ -4135,14 +4134,14 @@ Rossiya:Rossiya NP-TOP ; ! ""
 Kiev:Kiev NP-TOP ; ! ""
 Samarqand:Samarqand NP-TOP ; ! ""
 Qazaqstan:Qazaqstan NP-TOP ; ! ""
-O%'zbekstan:O%'zbekstan NP-TOP ; ! ""
-Qırg%'ızstan:Qırg%'ızstan NP-TOP ; ! ""
-O%'zbekstan:O%'zbekistan NP-TOP ; ! "" ! Dir/LR ! Use/Sub
-No%'kis:No%'kis NP-TOP ; ! ""
+ózbekstan:ózbekstan NP-TOP ; ! ""
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: capital

**Review by IlnarSelimcan — CHANGES_REQUESTED** [review]
@sharapat See inline comments .

.hfst files are binary files you get after compiling apertium-kaa. There's no need for tracking them in the repo. Please add the names of those files to `.gitignore` and remove them from the repo with `git rm`. If you know how to remove from the git history as well, please do so too :)

**Comment by IlnarSelimcan** [comment]
This is just for future reference:

## Coverage on New Testament (which is written in the older latin alphabet) before updating `kaa.lexc` to new alphabet:

```
apertium-kaa$ aq-covtest ../../../data4apertium/corpora/bible/kaa@latn.gospelgo.txt kaa.automorf.bin                                                                                         
Number of tokenised words in the corpus: 187417
Coverage: 87.89%
Top unknown words in the corpus:
835      a
616      an
520      Muxaddes
353      Masix
342      Menin
342      g
273      qa
260      atırg
199      Meni
183      Masixtın
178      ler
172      ın
166      ja
161      ha
161      Petr
159      Mag
148      Pavel
147      su
144      gu
143      mag
Translation time: 3.298170804977417 seconds
```

## Coverage on New Testament after updating `kaa.lexc`

```
apertium-kaa$ aq-covtest ~/1Working/1Taruen/apertiumpp/data4apertium/corpora/bible/kaa@latn.gospelgo.txt kaa.automorf.bin 
Number of tokenised words in the corpus: 245947
Coverage: 72.66%
Top unknown words in the corpus:
3464	 ha
2076	 m
1878	 ko
1568	 a
1102	 u
888	 ku
817	 Qudaydın
802	 so
781	 an
757	 z
733	 ju
679	 lar
668	 na
653	 ja
624	 gu
579	 tu
520	 Muxaddes
509	 su
504	 ler
501	 g
Translation time: 2.879375457763672 seconds
```

Coverage on NT should get up to 87.89% after kaa.twol is updated (and NT text is also updated and/or some compatibility mode added to `apertium-kaa`).

File: apertium-kaa.kaa.lexc [hunk 7]
```diff
@@ -4135,14 +4134,14 @@ Rossiya:Rossiya NP-TOP ; ! ""
 Kiev:Kiev NP-TOP ; ! ""
 Samarqand:Samarqand NP-TOP ; ! ""
 Qazaqstan:Qazaqstan NP-TOP ; ! ""
-O%'zbekstan:O%'zbekstan NP-TOP ; ! ""
-Qırg%'ızstan:Qırg%'ızstan NP-TOP ; ! ""
-O%'zbekstan:O%'zbekistan NP-TOP ; ! "" ! Dir/LR ! Use/Sub
-No%'kis:No%'kis NP-TOP ; ! ""
+Ózbekstan:Ózbekstan NP-TOP ; ! ""
+Qırǵızstan:Qırǵızstan NP-TOP ; ! ""
+ózbekstan:ózbekistan NP-TOP ; ! "" ! Dir/LR ! Use/Sub
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: Capitalize both ó.

File: apertium-kaa.kaa.lexc [hunk 8]
```diff
@@ -4246,7 +4245,7 @@ Berdaq:Berdaq NP-ANT-M ; ! ""
 Xristofor:Xristofor NP-ANT-M ; ! ""
 Yosif:Yosif NP-ANT-M ; ! ""
 Aleksandr:Aleksandr NP-ANT-M ; ! ""
-A%'bdiraman:A%'bdiraman NP-ANT-M ; ! ""
+ábdiraman:ábdiraman NP-ANT-M ; ! ""
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: бас әріп

File: apertium-kaa.kaa.lexc [hunk 9]
```diff
@@ -4337,36 +4336,36 @@ Arna:Arna NP-ANT-F ; ! "Arna" (Kazakh)
...
-A%'bil:A%'bil NP-ANT-M ; ! "Äbil" (Arabic)
-A%'bilqasım:A%'bilqasım NP-ANT-M ; ! "Äbilqasım" (Arabic)
-A%'bu:A%'bu NP-ANT-M ; ! "Äbiw" (Arabic)
-A%'di:A%'di NP-ANT-M ; ! "Ädiy" (Arabic)
-A%'dil:A%'dil NP-ANT-M ; ! "Ädil" (Arabic)
-A%'kim:A%'kim NP-ANT-M ; ! "Äkim" (Arabic)
-A%'lim:A%'lim NP-ANT-M ; ! "Älim" (Arabic)
-A%'mir:A%'mir NP-ANT-M ; ! "Ämir" (Kazakh)
+Ábdiraxman:Ábdiraxman NP-ANT-M ; ! "Äbdiraxman" (Arabic)
+ábil:ábil NP-ANT-M ; ! "Äbil" (Arabic)
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: 4340-4347 жолдар: бас әріп

File: apertium-kaa.kaa.lexc [hunk 10]
```diff
@@ -4380,43 +4379,43 @@ Esim:Esim NP-ANT-M ; ! "Esim" (Kazakh)
...
+Naǵım:Naǵım NP-ANT-M ; ! "Nağım" (Arabic)
 Narsha:Narsha NP-ANT-F ; ! "Narsha" (Kazakh)
-Na%'jip:Na%'jip NP-ANT-M ; ! "Näjip" (Arabic)
+Nájip:Nájip NP-ANT-M ; ! "Näjip" (Arabic)
 Niyaz:Niyaz NP-ANT-M ; ! "Nıyaz" (Arabic)
 Olga:Olga NP-ANT-F ; ! "OlʲGa" (Skandinavian)
 Oraz:Oraz NP-ANT-M ; ! "Oraz" (Old Turkic)
-O%'mir:O%'mir NP-ANT-M ; ! "Ömir" (Arabic)
-O%'ndiris:O%'ndiris NP-ANT-M ; ! "Öndiris" (Kazakh)
+ómir:ómir NP-ANT-M ; ! "Ömir" (Arabic)
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: бас әріп

File: apertium-kaa.kaa.lexc [hunk 11]
```diff
@@ -4380,43 +4379,43 @@ Esim:Esim NP-ANT-M ; ! "Esim" (Kazakh)
...
+óndiris:óndiris NP-ANT-M ; ! "Öndiris" (Kazakh)
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: бас әріп

File: apertium-kaa.kaa.lexc [hunk 12]
```diff
@@ -4537,7 +4536,7 @@ Lo:Lo NP-COG-MF ; ! ""
 van% Gog:van% Gog NP-COG-MF ; ! ""
 da% Vinchi:da% Vinchi NP-COG-MF ; ! ""
 Djobs:Djobs NP-COG-MF ; ! "Jobs"
-Tsukerberg:Tsukerberg NP-COG-MF ; ! "Zuckerberg"
+cukerberg:cukerberg NP-COG-MF ; ! "Zuckerberg"
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: бас әріп

File: apertium-kaa.kaa.lexc [hunk 13]
```diff
@@ -4583,7 +4582,7 @@ iPod:iPod NP-AL ; ! ""
 
 ! ************
 Abay:Abay NP-ANT-M ; ! ""
-A%'bdiraman:A%'bdiraman NP-ANT-M ; ! ""
+ábdiraman:ábdiraman NP-ANT-M ; ! ""
```
Comment (IlnarSelimcan) [inline apertium-kaa.kaa.lexc]: бас әріп.

And that's it. Everything else looks fine to me.

**Review by IlnarSelimcan — CHANGES_REQUESTED** [review]

File: apertium-kaa.kaa.lexc (hunk 10 above)
Comment (sharapat) [reply apertium-kaa.kaa.lexc]: Why ómir should start with capital letter?

Comment (jonorthwash) [reply apertium-kaa.kaa.lexc]: This is the name Ómir, not the common noun ómir.

**Comment by sharapat** [comment]
> ## Coverage on the [story](https://github.com/taruen/apertiumpp/blob/master/data4apertium/corpora/jam/kaa.txt) before @sharapat 's changes:
> ```
> apertium-kaa$ git log -1
> commit 54b29b3c1040fb45ead0155daa5ca3487ac934b5 (HEAD -> master, origin/master, origin/HEAD)
> Author: Ilnar Salimzianov <xxxxxxxxxxxxxx>
> Date:   Sun Apr 7 01:15:30 2019 +0300
> 
>     add .travis.yml and test.py
> 
> apertium-kaa$ aq-covtest ../../../data4apertium/corpora/jam/kaa.txt kaa.automorf.bin 
> Number of tokenised words in the corpus: 489
> Coverage: 69.53%
> Top unknown words in the corpus:
> 17	 Ayg
> 13	 l
> 13	 k
> 7	 j
> 5	 h
> 4	 redi
> 4	 ı
> 4	 shigi
> 3	 lken
> 3	 K
> 3	 ldi
> 3	 pıshı
> 2	 andı
> 2	 zir
> 2	 shik
> 2	 Yaq
> 2	 yde
> 2	 n
> 2	 m
> 2	 qayerge
> Translation time: 0.016431331634521484 seconds
> apertium-kaa$ 
> ```
> 
> ## Coverage on the story after sharapat's changes:
> ```
> apertium-kaa$ git log -1
> commit e7429c8f63c9cdb443ee0ffc0f8d897a64f47209 (HEAD -> coding-challenge, origin/coding-challenge)
> Author: sharapat <xxxxxxxxxxxxxxxxxxxxx>
> Date:   Wed Apr 24 00:26:03 2019 +0500
> 
>     changed sh -> ch where neccessary
> 
> apertium-kaa$ aq-covtest ~/1Working/1Taruen/apertiumpp/data4apertium/corpora/jam/kaa.txt kaa.automorf.bin 
> Number of tokenised words in the corpus: 413
> Coverage: 87.89%
> Top unknown words in the corpus:
> 4	 kóredi
> 3	 júda
> 3	 Aygúldiń
> 3	 pıshıǵı
> 2	 oynaǵandı
> 2	 Yaq
> 2	 úyde
> 2	 qayerge
> 1	 barqulla
> 1	 úydiń
> 1	 aldındaǵı
> 1	 oynaǵanına
> 1	 baratır
> 1	 Aygúlden
> 1	 kózlerin
> 1	 kórip
> 1	 turǵanı
> 1	 bolǵannan
> 1	 atirapına
> 1	 kórdińizbe
> Translation time: 0.019109010696411133 seconds
> ```
> 
> Unsurprisingly, the coverage has increased, which is a good thing.
> 
> @sharapat , this is how you can install aq-covtest: https://gist.github.com/IlnarSelimcan/87efbbe8a569833e68e89b126836e2bc . It is a tool for measuring the coverage of a transducer, i.e. it shows percentage of words in a corpus the transducer recognizes.
> 
> As the next step, please check whether the _stems_ of unknown words are in the `apertium-kaa.kaa.lexc`. In other words, make sure that `kór:kór V-TV ;` is there, and `Aygúl:Aygúl NP-ANT-F ;` is, etc.
> 
> If all stems from the story are already in the lexicon, it'll be easier for @jonorthwash to work on `apertium-kaa.kaa.twol` rules, since he'll know that unknown words `aq-covtest` returns from the story are unknown because of issues with twol, and _not_ because their stems are missing from `apertium-kaa.kaa.lexc`.
> 
> @sharapat , also, we'll need some more texts written in (or converted to) the new orthography (for testing purposes).
> 
> These can be Karakalpak texts which you'd like to see translated to Uzbek (personal interest in translations is a good motivator). Any text works, but ideally they should be under a free license (public domain, creative commons-by, creative-commons-by-sa).

Added the unknown words in apertium-kaa.kaa.lexc. Also edited atirapına to átirapına in [https://github.com/sharapat/story-translation-uzb-kaa-for-apertium/blob/master/kaa.txt](url), there was a mistake.

**Pull request merged**